File: dsa_leetcode_beginner/0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        int buyPrice = prices[0];
        int profit = 0;

        for (int i = 1; i < prices.size(); i++) {
            if (prices[i] < buyPrice) {
                buyPrice = prices[i];
            } else {
                int currentProfit = prices[i] - buyPrice;
                profit = std::max(profit, currentProfit);
            }
        }

        return profit;
    }
};

File: dsa_leetcode_beginner/0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # minimum buy price is the first price
        buy_price = prices[0]

        # the minimum profit is zero
        profit = 0

        for i in range(1, len(prices)):

            # if the current price is less, update the buy_price
            if prices[i] < buy_price:
                buy_price = prices[i]
            else:
                # else check if we can get a better profit
                current_profit = prices[i] - buy_price
                profit = max(current_profit, profit)

        return profit
    

#Question: https://leetcode.com/problems/best-time-to-buy-and-sell-stock
#Blog: https://blog.unwiredlearning.com/best-time-to-buy-and-sell-stock